### ISSUE TITLE
Visualizing the patch-level feature space

### ISSUE THREAD

**Issue by dsouzavijeth** [mahmoodlab/CLAM#53]
Hello. Thanks for your amazing work.

For visualizing patch-level feature space, in a subtyping task, it is said in the paper that "each patch is colored with the class for which the corresponding instance-level clustering network layer predicted positive (p ≥ 0.5). Patches predicted as negative (p < 0.5) for every single class is labeled as Agnostic. "

I would like to know if I'm understanding the steps correctly:

- patch features are extracted from the first fully-connected layer (out_features = 512)
- These features are input to the classifier (only the features (h), without attention score (A))
- logits from the classifier output are applied to a Softmax function that computes the predicted probability distribution over each class
- This output is used to classify the patch as one of the subtypes (p ≥ 0.5 & A_percentile ≥ 0.5) or Agnostic (p < 0.5 & A_percentile < 0.5)
- PCA is applied to reduce 512-dimensional patch-level feature space (h) to 2-dimensions and the corresponding predicted class is used to color the 2-d representation.

**Comment by fedshyvana** [comment]
Hi, patch-level feature visualization was done solely to interpret the results of the instance-level classifiers. So you're mostly correct, except for a n-class problem, the patch classification is done by passing each patch-level 512 dim features to each of the n instance-level classifiers (not the slide-level classifiers). You'll get n sets of 2-class prob scores [p0, p1] (after softmax), where p1 is the probability that the patch is positive for that class. The predicted class of the patch is the class for which p1 is >= 0.5, or if all p1 is <0.5, then it's called agnostic (we did not encounter patches for which p1 >= 0.5 for more than 1 class, but if that happens I suppose it's reasonable to pick the class with the highest score). 
Pca is performed to visualize features in 2d as you described.

**Comment by dsouzavijeth** [comment]
Yes, understood. Thank you for your detailed and informative response.
